plugins/reindex/src/main/java/org/elasticsearch/plugin/reindex/AbstractAsyncBulkByScrollAction.java

```diff
@@ -0,0 +1,352 @@
+                }
+            });
+        } catch (Throwable t) {
+            finishHim(t);
+        }
+    }
+
+    void onScrollResponse(SearchResponse searchResponse) {
+        threadPool.executor(ThreadPool.Names.GENERIC).execute(() -> {
+            try {
+                scroll.set(searchResponse.getScrollId());
+                SearchHit[] docs = searchResponse.getHits().getHits();
+                logger.debug("scroll returned [{}] documents with a scroll id of [{}]", docs.length, searchResponse.getScrollId());
+                if (docs.length == 0) {
+                    startNormalTermination();
+                    return;
+                }
+                batches.incrementAndGet();
+                List<SearchHit> docsIterable = Arrays.asList(docs);
+                if (mainRequest.size() != -1) {
+                    // Truncate the docs if we have more than the request size
+                    long remaining = max(0, mainRequest.size() - successfullyProcessed());
+                    if (remaining <= docs.length) {
+                        if (remaining < docs.length) {
+                            docsIterable = docsIterable.subList(0, (int) remaining);
+                        }
+                    }
+                }
+                BulkRequest request = buildBulk(docsIterable);
+                if (request.requests().isEmpty()) {
+                    /*
+                     * If we noop-ed the entire batch then just skip to the next
+                     * batch or the BulkRequest would fail validation.
+                     */
+                    startNextScrollRequest();
+                    return;
+                }
+                request.timeout(mainRequest.timeout());
+                request.consistencyLevel(mainRequest.consistency());
+                if (logger.isDebugEnabled()) {
+                    logger.debug("sending [{}] entry, [{}] bulk request", request.requests().size(),
+                            new ByteSizeValue(request.estimatedSizeInBytes()));
+                }
+                client.bulk(request, new ActionListener<BulkResponse>() {
+                    @Override
+                    public void onResponse(BulkResponse response) {
+                        onBulkResponse(response);
+                    }
+
+                    @Override
+                    public void onFailure(Throwable e) {
+                        finishHim(e);
+                    }
+                });
+            } catch (Throwable t) {
+                finishHim(t);
+            }
+        });
+    }
+
+    void onBulkResponse(BulkResponse response) {
+        try {
+            Set<String> destinationIndicesThisBatch = new HashSet<>();
+            for (BulkItemResponse item : response) {
+                if (item.isFailed()) {
+                    recordFailure(item.getFailure());
+                    continue;
+                }
+
+                switch (item.getOpType()) {
+                case "index":
```

Comment:
Hello from heathen land. ;) I thought the formatting rules on this would be a no-brainer as this is just another block statement but thanks for the explanation.